### PR TITLE
[Arc] Add arc resets/enables, mux-to-control-flow pass, and ZeroCount op

### PR DESCRIPTION
**Pull request by fabianschuiki**
Add a handful of improvements.

### CODE REVIEW

**Review by maerhart — APPROVED** [review]
I'm the author of these changes so I'm unable to provide a peer review of the code by definition. However, I'd be happy to see these changes merged 😄

File: include/circt/Dialect/Arc/CMakeLists.txt [hunk 1]
```diff
@@ -5,3 +5,8 @@ set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls)
 add_public_tablegen_target(CIRCTArcTransformsIncGen)
 add_circt_doc(Passes ArcPasses -gen-pass-doc)
+
+set(LLVM_TARGET_DEFINITIONS Arc.td)
+mlir_tablegen(ArcEnums.h.inc -gen-enum-decls)
+mlir_tablegen(ArcEnums.cpp.inc -gen-enum-defs)
+add_public_tablegen_target(CIRCTArcEnumsIncGen)
```
Comment (maerhart) [inline include/circt/Dialect/Arc/CMakeLists.txt:12]: We probably have to add it to the circt-headers as well here.
```suggestion
add_public_tablegen_target(CIRCTArcEnumsIncGen)
add_dependencies(circt-headers CIRCTArcEnumsIncGen)
```

**Comment by uenoku** [comment]
Note directly related to the PR but ZeroCountOp is really interesting. I guess Chisel emits mux-encoded ctz/clz for [priority encoder]( https://www.chisel-lang.org/api/3.2.8/chisel3/util/PriorityEncoder$.html). Perhaps in the future it might be good idea to create HW intrinsic for priority encoder so that we can try different very emission style (mux-encdoede, case and nested-if...).

**Comment by teqdruid** [comment]
> (Not directly related to PR) ZeroCountOp is really interesting! I guess Chisel emits mux-encoded ctz/clz for [priority encoder](https://www.chisel-lang.org/api/3.2.8/chisel3/util/PriorityEncoder$.html). Perhaps in the future it might be good idea to create HW intrinsic for priority encoder so that we can try different very emission style (mux-encdoede, case and nested-if...).

If we know that something is a priority encoder, we can optimize it based on the device/stdcell/compiler we're targeting. See my [LATTE'22 talk](https://capra.cs.cornell.edu/latte22/paper/10.pdf) ;)

File: include/circt/Dialect/Arc/Ops.td [hunk 2]
```diff
@@ -448,4 +452,20 @@ def LutOp : ArcOp<"lut", [
...
 
+def ZeroCountPredicateLeading  : I32EnumAttrCase<"leading", 0>;
+def ZeroCountPredicateTrailing  : I32EnumAttrCase<"trailing", 1>;
+def ZeroCountPredicate : I32EnumAttr<
+    "ZeroCountPredicate", "arc.zero_count predicate",
+    [ZeroCountPredicateLeading, ZeroCountPredicateTrailing]> {
+  let cppNamespace = "circt::arc";
+}
+
+def ZeroCountOp : ArcOp<"zero_count", [Pure, SameOperandsAndResultType]> {
```
Comment (darthscsi) [inline include/circt/Dialect/Arc/Ops.td:463]: LLVM is all over the place here.  CLZ and CTZ I think are separate, but all the comparisons are enumerated this way.  Just curious why you went with the enum.

**Review by darthscsi — APPROVED** [review]
Generally LGTM.

File: include/circt/Dialect/Arc/Ops.td (hunk 2 above)
Comment (fabianschuiki) [reply include/circt/Dialect/Arc/Ops.td:463]: @maerhart

Comment (maerhart) [reply include/circt/Dialect/Arc/Ops.td:463]: 1. CLZ and CTZ are very similar and having them in one op could help code sharing. E.g. canonicalization: one concrete example are the priority encoders which are a sequence of `comb.extract` and `comb.mux`, the extractions can go from high to low or the other way around. When this is not easy to detect immediately (e.g. when there is a function boundary between) we possibly have to reverse the order afterwards (as a canonicalization), we don't care about what the initial value of the enum is, we just want to flip it (this also saves us from building a new op and changing the use/def chain of the ssa values in that case).
2. I expect that most of the patterns/passes where we want to match on one of these, we want to match on both. That means, if we have separate ops we would need to do `isa<CLZ, CTZ>` or two separate dynamic casts, or define two separate patterns (possibly using templating) when using the dialect conversion/greedy rewrite framework or one that matches on generic `Operation`s.
3. That way it seems more consistend w.r.t. `comb.icmp` to me.

But I'd also be interested in your thoughts on this @darthscsi

**Pull request merged**